File: 2023/domaci/42-kuchynsky-robovysavac/volfmatous/main.js
const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false,
});

let linesRead = 0;
const lines = [];

async function readLine() {
    if (lines.length < 1) {
        for await (const line of rl) {
            lines.push(line);
        }
    }

    linesRead++;
    return lines[linesRead - 1];
}

const LEFT = 0;
const RIGHT = 1;
const FORWARD = 2;
const BACK = 3;

main().then();

async function main() {
    let inputSplit = (await readLine()).split(" ");
    const h = parseInt(inputSplit[0]);

    const floor = [];

    const tileColors = [];
    const program = new Map();
    for (let row = 0; row < h; row++) {
        floor.push((await readLine()).split(" "));

        for (const tile of floor[row]) {
            if (!program.has(tile)) {
                program.set(tile, LEFT);
                tileColors.push(tile);
            }
        }
    }

    inputSplit = (await readLine()).split(" ");
    let cleanerY = parseInt(inputSplit[0]) - 1;
    let cleanerX = parseInt(inputSplit[1]) - 1;

    inputSplit = (await readLine()).split(" ");
    let stationY = parseInt(inputSplit[0]) - 1;
    let stationX = parseInt(inputSplit[1]) - 1;

    let totalPossiblePrograms = 0;

    while (true) {
        if (tryProgram(floor, cleanerY, cleanerX, stationY, stationX, program)) {
            totalPossiblePrograms++;
        }

        let end = false;
        for (let i = 0; i < tileColors.length; i++) {
            program.set(tileColors[i], program.get(tileColors[i]) + 1);
            if (program.get(tileColors[i]) > 3) {
                if (i >= tileColors.length - 1) {
                    end = true;
                }

                program.set(tileColors[i], 0);
            } else {
                break;
            }
        }

        if (end) {
            break;
        }
    }

    console.log(totalPossiblePrograms);
}

function tryProgram(floor, cleanerY, cleanerX, stationY, stationX, program) {
    const floorVisited = [];
    for (let i = 0; i < floor.length; i++) {
        floorVisited.push([]);
        for (let j = 0; j < floor[0].length; j++) {
            floorVisited[i].push(false);
        }
    }

    while (true) {
        const tile = floor[cleanerY][cleanerX];
        const direction = program.get(tile);

        switch (direction) {
            case LEFT:
                cleanerX--;
                break;
            case RIGHT:
                cleanerX++;
                break;
            case FORWARD:
                cleanerY--;
                break;
            case BACK:
                cleanerY++;
                break;
        }

        if (cleanerY < 0 || cleanerY >= floor.length || cleanerX < 0 || cleanerX >= floor[0].length
            || floorVisited[cleanerY][cleanerX]) {
            return false;
        }

        if (cleanerY === stationY && cleanerX === stationX) {
            return true;
        }

        floorVisited[cleanerY][cleanerX] = true;
    }
}

File: 2023/domaci/42-kuchynsky-robovysavac/volfmatous/main.py
from typing import List, Dict

LEFT = 0
RIGHT = 1
FORWARD = 2
BACK = 3


def try_program(floor: List[List[str]], cleanerY: int, cleanerX: int, stationY: int, stationX: int,
                program: Dict[str, int]) -> bool:
    floor_visited = [[False] * len(floor[0]) for _ in range(len(floor))]

    while True:
        tile = floor[cleanerY][cleanerX]
        direction = program.get(tile, -1)

        if direction == LEFT:
            cleanerX -= 1
        elif direction == RIGHT:
            cleanerX += 1
        elif direction == FORWARD:
            cleanerY -= 1
        elif direction == BACK:
            cleanerY += 1

        if (cleanerY < 0 or cleanerY >= len(floor) or cleanerX < 0 or cleanerX >= len(floor[0])
                or floor_visited[cleanerY][cleanerX]):
            return False

        if cleanerY == stationY and cleanerX == stationX:
            return True

        floor_visited[cleanerY][cleanerX] = True


if __name__ == "__main__":
    inputSplit = input().split()
    h = int(inputSplit[0])

    floor = []
    tile_colors = []
    program = {}

    for row in range(h):
        floor.append([tile[0] for tile in input().split()])

        for tile in floor[row]:
            if tile[0] not in program:
                program[tile[0]] = LEFT
                tile_colors.append(tile[0])

    inputSplit = input().split()
    cleanerY = int(inputSplit[0]) - 1
    cleanerX = int(inputSplit[1]) - 1

    inputSplit = input().split()
    stationY = int(inputSplit[0]) - 1
    stationX = int(inputSplit[1]) - 1

    total_possible_programs = 0

    while True:
        if try_program(floor, cleanerY, cleanerX, stationY, stationX, program):
            total_possible_programs += 1

        end = False
        for tile_color in tile_colors:
            program[tile_color] += 1
            if program[tile_color] > 3:
                if tile_color == tile_colors[-1]:
                    end = True

                program[tile_color] = 0
            else:
                break

        if end:
            break

    print(total_possible_programs)
